### PR TITLE
fix multishop issues

### PR DESCRIPTION
**Pull request by jocel1**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | - images in the BO are now properly displayed depending on the shop context and in product admin page, fix switch from shop A to shop B where product doesn't exist
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Enable multishop, create a product in both shop A and shop B. Add an image in shop A, it should not appear in shop B. Create a product in shop A only. Edit the product, and switch to shop B. Information about the product should remain (only the images are not duplicated)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11028)

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/ProductController.php [hunk 1]
```diff
@@ -604,6 +611,48 @@ public function formAction($id, Request $request)
...
+        $this->duplicateProductCombinationsFromDefaultShop($product);
+
+        return $product;
+    }
+    /**
+     * Duplicate combinations from one shop to another
+     *
+     * @param Product $product
+     */
+    private function duplicateProductCombinationsFromDefaultShop(Product $product)
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/ProductController.php:633]: obviously we can't keep that here: object models calls are not allowed in the bundle: how about extract these 2 privates functions in a specific class with the responsability of duplication of a product?

File: src/PrestaShopBundle/Controller/Admin/ProductController.php [hunk 2]
```diff
@@ -604,6 +611,48 @@ public function formAction($id, Request $request)
...
+            foreach ($combinations as $id_combination) {
+                $combination = new Combination((int)$id_combination['id_product_attribute']);
+
+                $default_combination = new Combination(
+                    (int)$id_combination['id_product_attribute'],
+                    null,
+                    (int)$product->id_shop_default
+                );
+
+                $def = ObjectModel::getDefinition($default_combination);
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/ProductController.php:646]: it's a really low level call: shouldn't we add a new method to Combination instead so we keep the complexity in the model?

File: src/PrestaShopBundle/Controller/Admin/ProductController.php [hunk 3]
```diff
@@ -396,6 +400,9 @@ public function formAction($id, Request $request)
 
         $productAdapter = $this->get('prestashop.adapter.data_provider.product');
         $product = $productAdapter->getProduct($id);
+        if (Context::getContext()->shop->getContext() == Shop::CONTEXT_SHOP && !$product->isAssociatedToShop()) {
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/ProductController.php:403]: imo this is the responsibility of productAdapter to return the "right" new product.

As a proof, we can see a code smell: we affect $product variable twice.

File: src/PrestaShopBundle/Controller/Admin/ProductController.php [hunk 4]
```diff
@@ -60,6 +60,10 @@
 use PrestaShopBundle\Form\Admin\Product\ProductCategories;
 use Product;
 use Tools;
+use Shop;
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/ProductController.php:63]: We try to not rely on old models in PrestaShop bundle: in every minor release, we refactor this controller to remove piece by piece calls to legacy.

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]
Do you think we can add a unit | integration | e2e test for that? How can we check this fix the issue ?

File: src/PrestaShopBundle/Controller/Admin/ProductController.php (hunk 3 above)
Comment (jocel1) [reply src/PrestaShopBundle/Controller/Admin/ProductController.php:403]: That was also my first thought, but the ProductAdapter is called from several places, and I'm not sure we want this logic everywhere: it should applies only in the ProductController

File: src/PrestaShopBundle/Controller/Admin/ProductController.php (hunk 1 above)
Comment (jocel1) [reply src/PrestaShopBundle/Controller/Admin/ProductController.php:633]: agree, will do :)

File: src/PrestaShopBundle/Controller/Admin/ProductController.php [hunk 5]
```diff
@@ -604,6 +611,48 @@ public function formAction($id, Request $request)
...
+    /**
+     * Duplicate product info from one shop to another
+     *
+     * @param Product $product
+     * @throws \LogicException
+     */
+    private function duplicateProductFromDefaultShop(Product $product)
+    {
+        $productAdapter = $this->get('prestashop.adapter.data_provider.product');
+        $product = $productAdapter->getProduct($product->id, null, null, $product->id_shop_default);
```
Comment (123monsite-regis) [inline src/PrestaShopBundle/Controller/Admin/ProductController.php:623]: hi @jocel1 

 I test this pull on a 1.7.5.0, it works quite well  except that : 
after more tests `duplicateProductFromDefaultShop` and `$product = $productAdapter->getProduct($product->id, null, null, $product->id_shop_default);` is not corect, 
we may remplace this function by a function `duplicateProductFromTheShopWhereTheProductExist(..` as the product is not alawys duplicate from default shop but can come from another shop

i have not already idea about how to apply @mickaelandrieu reviwews who are justified ( how to respect Ps Standart , no Leagacy on the bundle )

regards

Comment (123monsite-regis) [reply src/PrestaShopBundle/Controller/Admin/ProductController.php:623]: ping @lagence123-dev

File: src/PrestaShopBundle/Controller/Admin/ProductController.php (hunk 1 above)
Comment (sarjon) [reply src/PrestaShopBundle/Controller/Admin/ProductController.php:633]: or even `DuplicateProductCommand` with its handler. :stuck_out_tongue:

**Comment by PierreRambaud** [comment]
Hi @jocel1 ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!
